flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala

```diff
@@ -1643,19 +1644,35 @@ object ScalarOperatorGens {
     val resultTypeTerm = primitiveTypeTermForType(componentInfo)
     val defaultTerm = primitiveDefaultValue(componentInfo)
 
+    if (index.literalValue.isDefined &&
+        index.literalValue.get.isInstanceOf[Int] &&
+        index.literalValue.get.asInstanceOf[Int] < 1) {
+      throw new ValidationException(s"Array element access needs an index starting at 1 but was " +
+        s"${index.literalValue.get.asInstanceOf[Int]}.")
+    }
     val idxStr = s"${index.resultTerm} - 1"
     val arrayIsNull = s"${array.resultTerm}.isNullAt($idxStr)"
     val arrayGet =
       rowFieldReadAccess(ctx, idxStr, array.resultTerm, componentInfo)
 
+    /**
+     * Return null when array index out of bounds which follows Calcite's behaviour.
+     * @see org.apache.calcite.sql.fun.SqlStdOperatorTable
+     */
     val arrayAccessCode =
-      s"""
-         |${array.code}
-         |${index.code}
-         |boolean $nullTerm = ${array.nullTerm} || ${index.nullTerm} || $arrayIsNull;
-         |$resultTypeTerm $resultTerm = $nullTerm ? $defaultTerm : $arrayGet;
-         |""".stripMargin
-
+    s"""
+        |${array.code}
+        |${index.code}
+        |$resultTypeTerm $resultTerm;
+        |boolean $nullTerm;
+        |if (${idxStr} < 0 || ${idxStr} >= ${array.resultTerm}.size()) {
```

Comment:
`${array.resultTerm}` maybe null?

Reply:
it seems `idxStr` could ` < 0` at runtime (if it's not a literal, and cannot validate before runtime), I think we can throw exception for this case.

Reply:
I think the `idxStr` has a certain value in runtime, both for `idx < 0` and 'idx >= array.size' we should return null.

Reply:
I think it's better to return null than throwing exception, because streaming is an online service. Besides, the definition of element accessing in Calcite also returns null for array element error. `org.apache.calcite.sql.fun.SqlStdOperatorTable#ITEM`

But could you add description for this on documentation? @leonardBang

Reply:
consider this case:
```sql
CREATE TABLE my_source (
  idx int,
  my_arr array<int>
) WITH (...);

SELECT my_arr[idx] FROM my_source;
```

Reply:
> consider this case:
> 
> ```sql
> CREATE TABLE my_source (
>   idx int,
>   my_arr array<int>
> ) WITH (...);
> 
> SELECT my_arr[idx] FROM my_source;
> ```
For this case, idx has been an integer value rather than an input reference in runtime IIUC.

Reply:
What I want to express is, the index could be a literal or a field reference.
And we want to throw Exception for index < 1, there are two cases that index < 1:
- literal value, it has been checked in code gen, and throws ValidationException.
- field reference, currently you just return null for this case.

Reply:
I got what's your concern @libenchao . I think we don't have divergency on the runtime null behavior, the only concern is the compile exception. But I still prefer the current way, because it's very likely a typo if `my_arr[0]` is used, so a hint exception is more useful than surprising result.